### ISSUE TITLE
0.35.0 - Name collisions

### ISSUE THREAD

**Issue by Benjamin-Dobell** [rollup/rollup#910]
I've got a project that uses Three.js (been building fine in several prior rollup versions) and it's now failing in a browser as at line 40907 is:

```
Object.defineProperties( Material.prototype, {
	wrapAround: {
		get: function () {
			console.warn( 'THREE.' + this.type + ': .wrapAround has been removed.' );
		},
		set: function ( value ) {
			console.warn( 'THREE.' + this.type + ': .wrapAround has been removed.' );
		}
	},
	wrapRGB: {
		get: function () {
			console.warn( 'THREE.' + this.type + ': .wrapRGB has been removed.' );
			return new Color();
		}
	}
} );
```

This is failing because `Material` was most recently defined as:

```
var Material = Object.freeze({
	ShadowMaterial: ShadowMaterial,
	SpriteMaterial: SpriteMaterial,
	RawShaderMaterial: RawShaderMaterial,
	ShaderMaterial: ShaderMaterial,
	PointsMaterial: PointsMaterial,
	MultiMaterial: MultiMaterial,
	MeshPhysicalMaterial: MeshPhysicalMaterial,
	MeshStandardMaterial: MeshStandardMaterial,
	MeshPhongMaterial: MeshPhongMaterial,
	MeshNormalMaterial: MeshNormalMaterial,
	MeshLambertMaterial: MeshLambertMaterial,
	MeshDepthMaterial: MeshDepthMaterial,
	MeshBasicMaterial: MeshBasicMaterial,
	LineDashedMaterial: LineDashedMaterial,
	LineBasicMaterial: LineBasicMaterial,
	Material: Material
});
```

Note that `Material` itself has a property referring to the earlier `Material` (a function).

It would seem this is meant to be a transpiled version of:

```
export { ShadowMaterial } from './ShadowMaterial.js';
export { SpriteMaterial } from './SpriteMaterial.js';
export { RawShaderMaterial } from './RawShaderMaterial.js';
export { ShaderMaterial } from './ShaderMaterial.js';
export { PointsMaterial } from './PointsMaterial.js';
export { MultiMaterial } from './MultiMaterial.js';
export { MeshPhysicalMaterial } from './MeshPhysicalMaterial.js';
export { MeshStandardMaterial } from './MeshStandardMaterial.js';
export { MeshPhongMaterial } from './MeshPhongMaterial.js';
export { MeshNormalMaterial } from './MeshNormalMaterial.js';
export { MeshLambertMaterial } from './MeshLambertMaterial.js';
export { MeshDepthMaterial } from './MeshDepthMaterial.js';
export { MeshBasicMaterial } from './MeshBasicMaterial.js';
export { LineDashedMaterial } from './LineDashedMaterial.js';
export { LineBasicMaterial } from './LineBasicMaterial.js';
export { Material } from './Material.js';
```

from `src/materials/material.js` - although I'm not sure how it ended up with the name `Material`.

0.35.0 seems like a big step in the right direction with regard to future-proofing and is great work in general (Thanks!). I'm so sorry to be dropping bug reports like this!

**Comment by Rich-Harris** [comment]
> I'm so sorry to be dropping bug reports like this!

Not at all, thanks for trying 0.35 out so soon and submitting these reports!

**Comment by nolanlawson** [comment]
I ran into this as well (https://github.com/pouchdb/pouchdb/pull/5673#issuecomment-246186805), and have a repro. I'll submit a test case and hopefully a fix if I have time.

**Comment by nolanlawson** [comment]
Unfortunately git bisect says the first bad commit is 173cfc0df75e84ed4303887741d18831ed0076c8 which is a lot of code 😕

**Comment by nolanlawson** [comment]
Good news! I have a fix. Will open a PR.

**Comment by Rich-Harris** [comment]
> which is a lot of code 😕

TFW when git bisect gives you the commit with a message that just says 'rewrite'...

Thanks @nolanlawson and @Benjamin-Dobell – have released 0.35.8 :metal:

**Comment by Benjamin-Dobell** [comment]
Hmm, unfortunately it doesn't look like #926 addressed the issue I'm seeing.

I'll try get a minimum repro.

**Comment by Benjamin-Dobell** [comment]
I'm not sure this is *exactly* issue I'm experiencing, as in my case `src/materials/material.js` is somehow ending up with the name `Material` even though it's only ever imported as `Materials` (plural). Nonetheless, it does demonstrate name collision behaviour between `import * as foo` and an pre-existing `foo`;

http://rollupjs.org/?shareable=JTdCJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmNqcyUyMiUyQyUyMm1vZHVsZU5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlMkMlMjJtb2R1bGVJZCUyMiUzQSUyMiUyMiU3RCUyQyUyMm1vZHVsZXMlMjIlM0ElNUIlN0IlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJpbXBvcnQlMjAqJTIwYXMlMjBDbGFzaGVzJTIwZnJvbSUyMCcuJTJGbGliX2VudHJ5LmpzJyUzQiU1Q24lNUNuQ2xhc2hlcy5NYXRlcmlhbCgpJTNCJTVDbkNsYXNoZXMuU29tZXRoaW5nKCklM0IlMjIlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIybWF0ZXJpYWxzLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMCU3QiUyME1hdGVyaWFsJTIwJTdEJTIwZnJvbSUyMCcuJTJGTWF0ZXJpYWwuanMnJTNCJTIyJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMk1hdGVyaWFsLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGZ1bmN0aW9uJTIwTWF0ZXJpYWwoKSUyMCU3QiU1Q24lMjAlMjBjb25zb2xlLmxvZyglNUMlMjJIZXkhJTVDJTIyKSUzQiU1Q24lN0QlMjIlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIyUHJvYmxlbS5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJpbXBvcnQlMjAlN0IlMjBNYXRlcmlhbCUyMCU3RCUyMGZyb20lMjAnLiUyRk1hdGVyaWFsLmpzJyUzQiU1Q24lNUNuT2JqZWN0LmRlZmluZVByb3BlcnRpZXMoTWF0ZXJpYWwucHJvdG90eXBlJTJDJTIwJTdCJTVDbiUyMCUyMCU1Q3R3cmFwQXJvdW5kJTNBJTIwJTdCJTVDbiU1Q3QlNUN0Z2V0JTNBJTIwZnVuY3Rpb24lMjAoKSUyMCU3QiU1Q24lNUN0JTVDdCU1Q3Rjb25zb2xlLndhcm4oJ1VoJTIwb2glMkMlMjB3ZSUyMHdhbnQlMjB0aGlzJTIwcHJvcGVydHklMjBkZWZpbmVkJTIwb24lMjBNYXRlcmlhbCUyMHRoZSUyMGZ1bmN0aW9uJTJDJTIwbm90JTIwdGhlJTIwbW9kdWxlJyklM0IlNUNuJTVDdCU1Q3QlN0QlNUNuJTVDdCU3RCU1Q24lN0QpJTNCJTIyJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMlNvbWV0aGluZy5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJpbXBvcnQlMjAqJTIwYXMlMjBNYXRlcmlhbCUyMGZyb20lMjAnLiUyRm1hdGVyaWFscy5qcyclM0IlNUNuJTVDbmV4cG9ydCUyMGZ1bmN0aW9uJTIwU29tZXRoaW5nKCklMjAlN0IlNUNuJTIwJTIwZm9yJTIwKHZhciUyMG0lMjBvZiUyME1hdGVyaWFsKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjBjb25zb2xlLmxvZygnU3R1ZmYnKSUzQiU1Q24lMjAlMjAlN0QlNUNuJTdEJTIyJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMmxpYl9lbnRyeS5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJleHBvcnQlMjAlN0IlMjBTb21ldGhpbmclMjAlN0QlMjBmcm9tJTIwJy4lMkZTb21ldGhpbmcuanMnJTNCJTVDbmV4cG9ydCUyMColMjBmcm9tJTIwJy4lMkZtYXRlcmlhbHMuanMnJTNCJTVDbmV4cG9ydCUyMColMjBmcm9tJTIwJy4lMkZQcm9ibGVtLmpzJyUzQiU1Q24lMjIlN0QlNUQlN0Q=

**Comment by Benjamin-Dobell** [comment]
@Rich-Harris Sorry, here's a much less convoluted [example](http://rollupjs.org/?shareable=JTdCJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmNqcyUyMiUyQyUyMm1vZHVsZU5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlMkMlMjJtb2R1bGVJZCUyMiUzQSUyMiUyMiU3RCUyQyUyMm1vZHVsZXMlMjIlM0ElNUIlN0IlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJpbXBvcnQlMjAlN0IlMjBTb21ldGhpbmclMjAlN0QlMjBmcm9tJTIwJy4lMkZTb21ldGhpbmcuanMnJTNCJTVDbmltcG9ydCUyMCU3QiUyME1hdGVyaWFsJTIwJTdEJTIwZnJvbSUyMCcuJTJGTWF0ZXJpYWwuanMnJTNCJTVDbiU1Q24lMkYlMkYlMjBUaGlzJTIwaXMlMjBnb2luZyUyMHRvJTIwY3Jhc2glMjBiZWNhdXNlJTIwd2UncmUlMjB0cnlpbmclMjB0byUyMGNhbGwlMjBPYmplY3QuZnJlZXplKCklMjBhcyUyMGElMjBtZXRob2QuJTVDbk1hdGVyaWFsKCklM0IlNUNuU29tZXRoaW5nKCklM0IlMjIlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIyTWF0ZXJpYWwuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyZXhwb3J0JTIwZnVuY3Rpb24lMjBNYXRlcmlhbCgpJTIwJTdCJTVDbiUyMCUyMGNvbnNvbGUubG9nKCU1QyUyMkhleSElNUMlMjIpJTNCJTVDbiU3RCUyMiU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJTb21ldGhpbmcuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyaW1wb3J0JTIwKiUyMGFzJTIwTWF0ZXJpYWwlMjBmcm9tJTIwJy4lMkZNYXRlcmlhbC5qcyclM0IlNUNuJTVDbmV4cG9ydCUyMGZ1bmN0aW9uJTIwU29tZXRoaW5nKCklMjAlN0IlNUNuJTIwJTIwZm9yJTIwKHZhciUyMG0lMjBvZiUyME1hdGVyaWFsKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjBjb25zb2xlLmxvZygnU3R1ZmYnKSUzQiU1Q24lMjAlMjAlN0QlNUNuJTdEJTIyJTdEJTVEJTdE) (I was working backwards!)

**Comment by nolanlawson** [comment]
@Benjamin-Dobell I'd recommend opening a new issue to start discussion. Sorry for misdiagnosing!

That example you post seems like a good repro. You can also add a test case to the code, which would be extremely helpful.

**Comment by nolanlawson** [comment]
@Benjamin-Dobell Thanks for the repro; found a fix ^

**Comment by Rich-Harris** [comment]
Have released 0.35.11 with the fix for the second bug – thanks everyone
